kvserver/rangefeed: process unregister requests asyncronously

stevendanna opened the pull request:
After the output loop completes, 2 cleanup steps are performed:

1) The registration is removed from the processor's registry, and
2) The process is potentially stopped and removed from the replica.

Before this change, both of these happened syncronously after the output loop finished.

With this change, step (1) happens asyncronously. To facilitate this, an overflow mechanism for requests is provided. This overflow mechanism potentially allocates. Note that we expect that the number of requests is relatively small and should be O(rangefeeds_on_range) so hopefully this mechanism won't be used often.

Step (2) is now handled by the processor itself. After processing an unregister request, if the set of registrations falls to zero, we enqueue a Stop event for ourselves. Then, when processing the Stop, we unregister ourselves from the replica.

Note that this may look like a small semantics change since the previous unregister callback called Stop() which processes all events. However, note that a Stopped event is processed after all other events, so any events in the queue at the point of processing the unregistration that enqueued the stop will be processed.

The motivation for this change is to eventually allow cleanup step (1) to be run as part of registration.disconnect(), which needs to be non-blocking. This is desired for a future change in which there is not a dedicated goroutine to perform this cleanup.

Epic: none
Release note: None

stevendanna committed:
kvserver/rangefeed: process unregister requests asyncronously

After the output loop completes, 2 cleanup steps are performed:

1) The registration is removed from the processors registry
2) The process is potentially stopped and removed from the replica.

Before this change, both of these happened syncronously after the
output loop finished.

With this change, step (1) happens asyncronously. To facilitate this,
an overflow mechanism is provided. This overflow mechanism potentially
allocates. Note that we expect that the number of requests is
relatively small and should be O(rangefeeds_on_range) so hopefully
this mechanism won't be used often.

Step (2) is now handled by the processor itself. After processing an
unregister request, if the set of registrations falls to zero, we
enqueue a Stop event for ourselves. Then, when processing the Stop, we
unregister ourselves from the replica.

Note that this may look like a small semantics change since the
previous unregister callback called Stop() which processes all events.
However, note that a Stopped event is processed after all other
events, so any events in the queue at the point of processing the
unregistration that enqueued the stop will be processed.

The motivation for this change is to eventually allow cleanup step (1)
to be run as part of registration.disconnect(), which needs to be
non-blocking. This is desired for a future change in which there is
not a dedicated goroutine to perform this cleanup.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132857)
<!-- Reviewable:end -->

wenyihu6 commented on pkg/kv/kvserver/rangefeed/scheduled_processor.go:
```diff
@@ -346,13 +395,7 @@ func (p *ScheduledProcessor) Register(
-					f()
-				}
-			}
+			p.unregisterClientAsync(r)
```
Still reading on this PR more, but QQ: Can we move this to r.disconnect now?

wenyihu6 reviewed (commented):
Only took a look at the first commit atm - this is going to make the cleanup process a lot smoother. I’ve left a few questions where I wasn’t entirely sure. Lmk if you would prefer a sync call to align on this PR. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-:-O9Ug93x2xLKzfK-zXn_:b-iegnwl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)
 
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 351 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9UNjvk8bAMFJph1n0N:-O9UNjvk8bAMFJph1n0O:befgw79) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/registry.go#L351)):*
> ```Go
> 	reg.metrics.RangeFeedRegistrations.Dec(int64(reg.tree.Len()))
> 	reg.forOverlappingRegs(ctx, all, func(r registration) (bool, *kvpb.Error) {
> 		r.drainAllocations(ctx)
> ```

1. Do we need to move r.drainAllocations inside the forOverlappingRegs? I'm worried about a race condition could be possible here - what if events came in after draining? So r.disconnect must happen before r.drainAllocations?

```
		dis, pErr := fn(r)
		if dis {
			// could we do r.drain here
			r.disconnect(pErr)
			r.drainAllocations(ctx)
			toDelete = append(toDelete, i)
		}
```

2. To make sure I'm aligned with your understanding, is it correct that we are doing r.drainAllocations separately for reg.DisconnectAll because we will be 
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ufa8ODSWdtiBaHreD:-O9Ufa8PBQOhEUjmYrC5:biau0xa) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L197)):*
> ```Go
> 
> 	toDrain := p.overflowedUnregRequestQueue.reqs
> 	p.overflowedUnregRequestQueue.reqs = make([]request, 0, 8)
> ```

For my education, you mentioned that having the fixed chunk size queue PR would be helpful here. Could you help me better understand what the performance implications on why that is the case?
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 274 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Uaf2BEh9oKieUhn5e:-O9Uaf2BEh9oKieUhn5f:bpcvop3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59f52bc6d9ec3a6e5dbe8a333c29bff5a2c0336/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L274)):*
> ```Go
> func (p *ScheduledProcessor) sendStop(pErr *kvpb.Error) {
> 	p.enqueueRequest(func(ctx context.Context) {
> 		p.reg.DisconnectWithErr(ctx, all, pErr)
> ```

Question about existing code - is the reason we are Disconnecting al here and again when we are processing the stop request during 
```
func (p *ScheduledProcessor) processStop() {
	p.cleanup() // also calls p.reg.DisconnectAllOnShutdown(ctx, pErr)
```
that we want to disconnect as early as possible but there may be new things coming after this since p.stopping is not protected under a mutex?
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 277 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ucdtf6UrUrPpKksqi:-O9Ucdtf6UrUrPpKksqj:bqtkn6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59f52bc6d9ec3a6e5dbe8a333c29bff5a2c0336/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L277)):*
> ```Go
> 		// First set stopping flag to ensure that once all registrations are removed
> 		// processor should stop.
> 		p.stopping = true
> ```

Question about existing code - do we need to protect this stopping flag under a mutex and move DisconnectWithError (all) below this line? What if new p.Register came in before this and request succeeds since p.stopping is false? But I guess we will be shutting down again during p.processStop()? 
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ubf2U9AUdAQ7gkxpB:-O9Ubf2U9AUdAQ7gkxpC:b-9xi6hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L321)):*
> ```Go
> func (p *ScheduledProcessor) stopInternal(ctx context.Context, pErr *kvpb.Error) {
> 	p.overflowedUnregRequestQueue.Lock()
> 	p.overflowedUnregRequestQueue.ignoreAdditionalUnregRequests = true
> ```

Why is this necessary? Is it because we do not want to enqueue any unreg requests during shutdown since those requests won't do anything anyways? 
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 433 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ucl637EmsbifwEuC3:-O9Ucl637EmsbifwEuC4:bj8oymx) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L433)):*
> ```Go
> 		// processor.
> 		if p.reg.Len() == 0 {
> 			p.stopInternal(ctx, nil)
> ```

1. The old code on `replica.maybeDisconnectEmptyRangefeed()` calls `p.syncEventC()` before sending stop, and then unsetRangefeedProcessor. Do we need to call `p.syncEventC()`  somewhere as well?
2. Do we also need to check for `!r.updateRangefeedFilterLocked()` like what `replica.maybeDisconnectEmptyRangefeed()` had?
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 713 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9UdDlG5vdpFdfKA0pk:-O9UdDlG5vdpFdfKA0pl:b-3zizlf) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L713)):*
> ```Go
> 		p.overflowedUnregRequestQueue.Lock()
> 		defer p.overflowedUnregRequestQueue.Unlock()
> 		if p.overflowedUnregRequestQueue.ignoreAdditionalUnregRequests {
> ```

Are there any reasons why we don't return early if `ignoreAdditionalUnregRequests` is set before selecting on a channel and sending request here?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 351 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9UNjvk8bAMFJph1n0N:-O9UhRku9ICa1HZz_UDa:bd4vhib) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/registry.go#L351)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

1. Do you think we can move r.drainAllocations inside the forOverlappingRegs? Do you think a race condition could be possible here - what if events came in after draining? So r.disconnect must happen before r.drainAllocations?

```
		dis, pErr := fn(r)
		if dis {
			// could we do r.drain here
			r.disconnect(pErr)
			r.drainAllocations(ctx)
			toDelete = append(toDelete, i)
		}
```

2. To make sure I'm aligned with your understanding, is it correct that we are doing r.drainAllocations separately for reg.DisconnectAll because we will be
</blockquote></details>

1. My belief is that there is no race because DisconnectAllOnShutdown is called during processor shutdown. The processor is what delivers messages to the registration, so we know it won't be getting any more message.
2. My belief is that in the worse case, we are calling this twice for some registrations. But that seems ok, if we called it before, it should be fast.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 398 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132857#gh-1805219571:-O9Ui-KOElhafdtEte0N:blxvkde) ([raw file](https://github.com/cockroachdb/cockroach/blob/91dc0275d4782d49dc196983ee88e576024405a9/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L398)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Giving this PR a closer look, but QQ: Can we move this to r.disconnect now? 
</blockquote></details>

This is done in a follow up commit
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ufa8ODSWdtiBaHreD:-O9Ui57e1gBcEJqGG1_d:bn6s376) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L197)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

For my education, you mentioned that having the fixed chunk size queue PR would be helpful here. Could you help me better understand what the performance implications on why that is the case?
</blockquote></details>

Reviewable has attached this to some unrelated code to me. But, before I had moved this to a simple bool, I had a slice as a fallback queue.  In that case, I thought it might be nice to have a queue that amortized allocations a bit better than a slice may, but I hadn't thought about it very hard. In the end, I decided that there are lots of cases where we might effectively end up iterating every registration already (for instance, sending a message that every registration is subscribed to), so it wouldn't be too harmful to do that in the overflow case in genera and avoid keeping the overflowed events around.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 274 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Uaf2BEh9oKieUhn5e:-O9UnOG4E1Pp5H5o_tl0:b-f3ohgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59f52bc6d9ec3a6e5dbe8a333c29bff5a2c0336/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L274)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Question about existing code - is the reason we are Disconnecting al here and again when we are processing the stop request during 
```
func (p *ScheduledProcessor) processStop() {
	p.cleanup() // also calls p.reg.DisconnectAllOnShutdown(ctx, pErr)
```
that we want to disconnect as early as possible but there may be new things coming after this since p.stopping is not protected under a mutex?
</blockquote></details>

I think the main thing is that we want to deliver the correct error. Anyone that is still around by the time we call DisconnectAllOnShutdown gets a potentially different error.  As I mentioned elsewhere, requests are processed one at a time in the event loop. A registration is also a request, so by the time stopInternal returns we know both that all nodes have been disconnected and that no new registrations are allowed.  We have the extra DisconnectAllOnShutdown in cleanup() for the case of the scheduler enqueing a Stopped event because of a whole processor stop, which won't go through this code.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 277 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ucdtf6UrUrPpKksqi:-O9UksF-9n21H69vk9jW:b-deqzfm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59f52bc6d9ec3a6e5dbe8a333c29bff5a2c0336/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L277)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Question about existing code - do we need to protect this stopping flag under a mutex and move DisconnectWithError (all) below this line? What if new p.Register came in before this and request succeeds since p.stopping is false? But I guess we will be shutting down again during p.processStop()?
</blockquote></details>

A thing to remember is that what makes most of scheduled processor thread safe is that we only update the processor in the request loop avoiding a mutex in most cases.  I added a mutex (eventually changed to an atomic) for the overflow queue, because we explicitly want to take an action outside of the event loop in that on ecase.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ubf2U9AUdAQ7gkxpB:-O9UlBRW2KpZ6dVzqqJ4:b-ig5jpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L321)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Why is this necessary? Is it because we do not want to enqueue any unreg requests during shutdown since those requests won't do anything anyways?
</blockquote></details>

This was removed in a follow-up, but my idea here was that I didn't want shutdown, which might call a flurry of disconnects, to allocate a bunch of things in the overflow queue when we knew we were about to tear down everything anyway.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 433 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9Ucl637EmsbifwEuC3:-O9UlRc_1OET2Ktwlu4k:b26aiz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L433)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

1. The old code on `replica.maybeDisconnectEmptyRangefeed()` calls `p.syncEventC()` before sending stop, and then unsetRangefeedProcessor. Do we need to call `p.syncEventC()`  somewhere as well?
2. Do we also need to check for `!r.updateRangefeedFilterLocked()` like what `replica.maybeDisconnectEmptyRangefeed()` had?
</blockquote></details>

1. I mentioned this in the commit but I should have been more explicit.  stopInternal "enqueues" a Stopped event at the _scheduler_ level. In process() we handled Stopped events last, after we process any events in the event queu. So we know that any events in the queue at the point of doing this will be processed before we process the stop because of the order of execution in process:

```
func (p *ScheduledProcessor) process(e processorEventType) processorEventType {
	ctx := p.processCtx
	if e&RequestQueued != 0 {
		p.processRequests(ctx)
	}
	if e&EventQueued != 0 {
		p.processEvents(ctx)
	}
	if e&PushTxnQueued != 0 {
		p.processPushTxn(ctx)
	}
	if e&Stopped != 0 {
		p.processStop()
	}
	return 0
}
```

We are current in `processRequests`, so `processEvents` _will_ happen if there are any events, before we get stopped.


2. I don't _think_ so. Namely, as far as I can tell, the only way updateRangefeedFilterLocked() was ever going to return `false` was if the processor was stopped. In my change, we are now unregistering ourselves as part of the processor stopping process, so it should be the same.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 713 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9UdDlG5vdpFdfKA0pk:-O9Un0uu1U02Lkb01Lk9:bhxumm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L713)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Are there any reasons why we don't return early if `ignoreAdditionalUnregRequests` is set before selecting on a channel and sending request here?
</blockquote></details>

If this code were still around it would be a good point. In general I wanted to avoid touching this lock whenever I didn't need to.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
I’ll reach out on Slack with some of my uncertainty and resolve the comments. Overall, I think this is a solid approach and I can see it making through - I just need to convince myself a bit more since I'm not super familiar with the scheduled processor code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-:-O9V0r3yDY0u_lx2ap-f:bsn82ww)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

stevendanna commented:
@nvanbenschoten Wondering what you think about the general direction here. Note that if you review commit-by-commit, the second commit changes how overflow is handled substantially.

stevendanna commented on pkg/kv/kvserver/rangefeed/scheduled_processor.go:
```diff
@@ -633,6 +654,33 @@ func (p *ScheduledProcessor) enqueueRequest(req request) {
+	case p.requestQueue <- req:
+		p.scheduler.Enqueue(RequestQueued)
+	default:
+		r.setShouldUnregister(true)
+		p.overflowedUnregRequests.Store(true)
+		p.scheduler.Enqueue(RequestQueued)
```
We can make this the _only_ way unregister requests are handled and simplify this code further. Then, all the tests for unregistration are also hitting this path.

tbg replied:
Also, are we messing up the metrics update if we take this path? I don't see us call `RangeFeedRegistrations.Dec(.)` in `unregisterMarkedRegistrations`.
<!-- Reviewable comment -OBPbAux06TPs8tNLqmQ:b-bub0q1 -->
<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
Looks good so far. I see that there are merge commits, and we have the outstanding TODO about unifying the unregistration path. Let me know when it's ready for another look!

> Reviewed 2 of 4 files at r1, 2 of 2 files at r3, 3 of 3 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-:-OBPdDf7-b5SMxztYX4d:b65jce8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @stevendanna, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 351 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9UNjvk8bAMFJph1n0N:-OBPOPu57M4qIbUu0D73:b-rb7z94) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/registry.go#L351)):*
> My belief is that in the worse case, we are calling this twice for some registrations. But that seems ok, if we called it before, it should be fast.

Wouldn't that then lead to an incorrect `RangeFeedRegistrations` metric? That would seem problematic. We should be able to guarantee that `DisconnectAllOnShutdown` is called exactly once, or make it idempotent.
___
*[`pkg/kv/kvserver/rangefeed/buffered_registration.go` line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OBPQnb41G17225gfvES:-OBPQnb41G17225gfvET:braw7ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c7f4f377c597953f6a3f1197a290cd6056e1c68/pkg/kv/kvserver/rangefeed/buffered_registration.go#L83)):*
> ```Go
> 	metrics *Metrics,
> 	stream Stream,
> 	unregisterFn func(),
> ```

`unregister` and `cleanup` are hard to interpret without reading code (as I now have). Can you add comments on these here (just say what is passed when and delineate the responsibilities of `unregister` and `cleanup`, including a clean contract on when they are called and how they tie in with `br.disconnect`). The fields on `baseRegistration` should then also refer to this comment (or vice versa).

I think `cleanup` is really `(*Replica).maybeDisconnectEmptyRangefeed` in both the scheduled and legacy processor world. `unreg` is really `(*ScheduledProcessor).unregisterClientAsync` in the new world, and a no-op in the old world.

The current names seem to obfuscate that unnecessarily, if we literally renamed `cleanup` to `replicaMaybeDisconnectEmptyRangefeed` it would be clearer (similar comment for `unregisterFn`).


<!-- Sent from Reviewable.io -->

stevendanna committed:
kvserver/rangefeed: remove unused unreg function

The `unreg` function was only used by the LegacyProcessor, which no
longer exists.

Here, we remove it, rename `cleanup` to a more descriptive
name, and remove an unused context argument.

Epic: none
Release note: None

stevendanna committed:
kvserver/rangefeed: always use async unregister path

The number of registrations for a given processor is typically on the
order of the number of rangefeeds running against the range.  Further,
we already have many operations that require scanning the entire
registry. As a result, we can simplify the unregistration code by
always using what was previously the "overflow" path.

I've also removed some functions that were no or already were only
called in tests.

stevendanna committed:
kvserver/rangefeed: clarify registration metric update

We were previously updating this metric in different functions,
leading to doubt over whether our updates were correct. Here, we move
the metric decrement to the only function that is deleting anything
from the registry tree.  This is symetric with where we increment the
metric.

Epic: none
Release note: None

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @tbg, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 351 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-O9UNjvk8bAMFJph1n0N:-OBQ8KxO8JdTJ_aBK-zb:b-8l2kv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/246f79dde01870338fa0b5abfcc45c5b101d65ee/pkg/kv/kvserver/rangefeed/registry.go#L351)):*
The 'this' in that sentence was referring to drainAllocations() which wouldn't impact the metric.  DisconnectAllOnShutdown is fine to call more than once with respect to the metric because we only decrement by the length of the tree before removing everything from the tree. If it got called again, anything in the tree must be newly added.  I clarified this in the code a bit by moving the metric decrement to the only code location where we Delete() from the tree so it is symmetric with the metric increment when we Insert. I think the subtly here is that it is hard to notice that the `true` returned by this callback is really `disconnectAndRemoveFromTree`.

> Do you think we can move r.drainAllocations inside the forOverlappingRegs? Do you think a race condition could be possible here what if events came in after draining? So r.disconnect must happen before r.drainAllocations?

@wenyihu6 Coming back to this now that we have some space: I think my argument here was maybe wrong for some code paths. But, I have a follow up question: could we do this after runOutputLoop returns instead?  Then, in your larger PR, this would actually be right where we drain the catchup buffer for unbuffered registration, so it would all line up nicely. I think I'm going to push a small PR that does this. 


<!-- Sent from Reviewable.io -->

stevendanna committed:
rangefeed: remove minor duplication

Epic: none
Release note: None

stevendanna commented:
@tbg @wenyihu6 I've rebased this against all the other work merged to master and re-arranged some of the metrics code to make it more clear where things are being accounted for.

tbg reviewed (commented):
Need a second pass, first pass mostly for the broad strokes.
I reviewed commit-by-commit and so left some comments that are obsolete by the end. Please just dismiss them :-)

> Reviewed 1 of 9 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-:-OCcudfP80jY2kt4OidH:bjjw9tz)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @stevendanna, and @wenyihu6)

___
*[`-- commits` line 7 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcpfLs32pIk5TjbdBa:-OCcpfLs32pIk5TjbdBb:b9ieqpk):*
nit: processor
___
*[`-- commits` line 15 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcpmmpCuN41IHWMJrG:-OCcpmmpCuN41IHWMJrH:bpxp9oy):*
just checking that this is still true.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 74 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcqbQ565zgkgdWGMXq:-OCcqbQ565zgkgdWGMXr:b-yap2rf) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/registry.go#L74)):*
> ```Go
> 	getUnreg() func()
> 
> 	// shouldUnregister returns true if this registration should
> ```

You probably don't like that we have a random setter/getter pair for a bool on here either. Is this considered an interim step? I don't remember a Go koan that says you should communicate data through interfaces :smirk: 
Since I assume we'll live with this for now, comments with some explanatory power would be nice. I tried to suggest a comment but realized I have to figure out how this works from code first, so maybe I'll wait for the comments and then spend less time figuring it out since I know what I'll be looking for :laughing: 
In particular, who calls `shouldUnregister`? It can't be the registration, since the registration doesn't have to go through the interface for itself. Looking at the code, I see that it is the registry. Ok. Who calls `shouldUnregister`? Ah, the `Processor`, since it does the unregistration async (as the commit msg explains, though I'm not sure comments do). All of these should be answered by the comments to make life easy for the next Tobi.
___
*[`pkg/kv/kvserver/rangefeed/processor.go` line 111 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcqNUT3pmAe-1w6tJb:-OCcqNUT3pmAe-1w6tJc:b95ox3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/processor.go#L111)):*
> ```Go
> 	// replica that this processor can call when shutting down to
> 	// remove itself from the replica.
> 	UnregisterFromReplica func(Processor)
> ```

a Replica can have multiple processors? Or why is there a need to pass the processor in? I thought it was attached to the Replica (I might be wrong).
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 408 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcuSzx74gBIufAV96E:-OCcuSzx74gBIufAV96F:bgc7vqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3deda1a3b0f4f90fb5ea44689cecccf18815acc2/pkg/kv/kvserver/rangefeed/registry.go#L408)):*
> ```Go
> 	// registry, so we can assume we can decrement this by the
> 	// lenght of the input.
> 	reg.metrics.RangeFeedRegistrations.Dec(int64(len(toDelete)))
> ```

:heart: 
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 169 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCctrWR45bfuYd70CDZ:-OCctrWR45bfuYd70CD_:by7iyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6898797e5784a8d3007a6712666f39c5eaf0c6dd/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L169)):*
> ```Go
> 				p.reg.unregisterMarkedRegistrations(ctx)
> 				// If we have no more registrations, we can stop this
> 				// processor.
> ```

You could add a sentence explaining why no new registrations can have come in in the meantime.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 644 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcu1e1APAuXBrMR20o:-OCcu1e1APAuXBrMR20p:b-l1tzc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6898797e5784a8d3007a6712666f39c5eaf0c6dd/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L644)):*
> ```Go
> 
> // unregisterClientAsync instructs the processor to scan the registration for
> // clients to unregister on the next request. This doesn't send and actual
> ```

nit: any actual
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 656 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcu9SX24JGtdGVa-Gg:-OCcu9SX24JGtdGVa-Gh:b-4coroi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6898797e5784a8d3007a6712666f39c5eaf0c6dd/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L656)):*
> ```Go
> 	default:
> 	}
> 	r.setShouldUnregister(true)
> ```

this is never called with `false`, right? Can you simplify the interface to that extent? It will be helpful not to have to think about the possibility of anyone calling `setShouldUnregister(false)`. (`setShouldUnregister()`).


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @tbg, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/processor.go` line 111 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcqNUT3pmAe-1w6tJb:-OCd4B7G9lS4bKnlj7zk:b-1imyl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/processor.go#L111)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

a Replica can have multiple processors? Or why is there a need to pass the processor in? I thought it was attached to the Replica (I might be wrong).
</blockquote></details>

The function takes a Processor since the implementation code checks that the processor you are trying to remove is the same as the one stored on the replica.  Even with my change here, there are still a few different places where this disconnection happened so I didn't want to remove the check. 
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 74 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcqbQ565zgkgdWGMXq:-OCd4ty133eo0ZjVqbqC:b-p4cvol) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/registry.go#L74)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You probably don't like that we have a random setter/getter pair for a bool on here either. Is this considered an interim step? I don't remember a Go koan that says you should communicate data through interfaces :smirk: 
Since I assume we'll live with this for now, comments with some explanatory power would be nice. I tried to suggest a comment but realized I have to figure out how this works from code first, so maybe I'll wait for the comments and then spend less time figuring it out since I know what I'll be looking for :laughing: 
In particular, who calls `shouldUnregister`? It can't be the registration, since the registration doesn't have to go through the interface for itself. Looking at the code, I see that it is the registry. Ok. Who calls `shouldUnregister`? Ah, the `Processor`, since it does the unregistration async (as the commit msg explains, though I'm not sure comments do). All of these should be answered by the comments to make life easy for the next Tobi.
</blockquote></details>

Thanks for the nudge, I was leaving this despite disliking it. I _think_ we can make this a bit cleaner by reusing IsDisconnected now that we have that method on our interface here anyway.  It's always reasonable to unregister disconnected registrations since they won't be processing any of our messages anyway. 


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @tbg, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 74 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcqbQ565zgkgdWGMXq:-OCd6miv3dd46UfXrYQ5:b-um20ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/registry.go#L74)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Thanks for the nudge, I was leaving this despite disliking it. I _think_ we can make this a bit cleaner by reusing IsDisconnected now that we have that method on our interface here anyway.  It's always reasonable to unregister disconnected registrations since they won't be processing any of our messages anyway.
</blockquote></details>

Although, I think the reason I didn't do that is to avoid having to think about the complex locking situation too much when reading this code. I'll poke at it for a bit and see which is worse.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> and thanks for following through! Some more requests for comments but on the substance I see what's going on here and it seems to check out. Please let @wenyihu6 get a good look at this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-:-OCdBHdQDub0lD-FExGd:b-hacx8z)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten, @stevendanna, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 74 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcqbQ565zgkgdWGMXq:-OCd7ULc77egYDSWC9cz:b6kdcld) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/registry.go#L74)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Although, I think the reason I didn't do that is to avoid having to think about the complex locking situation too much when reading this code. I'll poke at it for a bit and see which is worse.
</blockquote></details>

Maybe for a follow-up PR?
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 91 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCd84LT3kokmnwEcVfH:-OCd84LU8m60amW32yav:bs60hea) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/registry.go#L91)):*
> ```Go
> 	// during disconnect(). Since it is called during disconnect it must be
> 	// non-blocking.
> 	removeRegFromProcessor func(registration)
> ```

this can be a `func()` (I understand that you need to pass a `func(reg)` into `newBufferedRegistration` but you can closure-wrap it  after that point).
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 96 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCd8GP0BVpgYdvwR7TG:-OCd8GP0BVpgYdvwR7TH:b-jrxcue) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/registry.go#L96)):*
> ```Go
> 	id               int64         // internal
> 	keys             interval.Range
> 	shouldUnreg      atomic.Bool
> ```

comment
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 423 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCd8eB941c5iQ5BlzHL:-OCd8eB941c5iQ5BlzHM:bevcwiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/registry.go#L423)):*
> ```Go
> }
> 
> func (reg *registry) unregisterMarkedRegistrations(ctx context.Context) {
> ```

comment (incl. that this is called by the processor). Or you could just reference some "master comment" on the unregistration sequence that exists elsewhere.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 237 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCd8xLo2s0iHdYM9czM:-OCd8xLo2s0iHdYM9czN:bq2avzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L237)):*
> ```Go
> }
> 
> func (p *ScheduledProcessor) cleanup() {
> ```

I don't know if this is comment worthy, but I went to read the code and this is called in two places:

- when the stopper starts draining before we start the processor task. In that case, we just call cleanup but don't have to update the metric since we never started the proc.
- the normal path, through `processStop` triggered by the proc's main handling loop. In that case, caller decrements the metric.

But there's also `stopInternal`. This is called from the main event loop, and sits a level higher in the machine. It disconnects all registrations, marks itself as stopping, and schedules a Stop event to be sent to itself. That stop event will then eventually trigger `processStop` above. Man, that needs to be spelled out somewhere and then referenced repeatedly in comments all over the place.

I wish there were a way to make it clear, just by looking at a method, who calls it (is this on the scheduler goroutine? Is it an ad-hoc call? How many levels deep are we here?) but I don't think that's a yak worth shaving in this PR.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 256 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCdAMDi3YEvTdXQ0X01:-OCdAMDi3YEvTdXQ0X02:bgj1pok) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L256)):*
> ```Go
> 	close(p.stoppedC)
> 	p.MemBudget.Close(ctx)
> 	if p.UnregisterFromReplica != nil {
> ```

when would this be nil?  Can it never be nil please? The field comment doesn't tell me when it is nil.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 649 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCdAyRw5L0dtdFvE1fQ:-OCdAyRw5L0dtdFvE1fR:b-57679p) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L649)):*
> ```Go
> // We are OK with these being processed out of order since these requests
> // originate from a registration cleanup, so the registration in question is no
> // longer processing event.
> ```

nit: events


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
 
I think I've convinced myself in the correctness of this PR. I feel like some aspects of the shutdown flow could use cleanup. We can tackle them as follow-ups in https://github.com/cockroachdb/cockroach/issues/135332.

1. Do we still need this https://github.com/cockroachdb/cockroach/blob/95584215c8a48bc5e317fae189f6d0ed70795360/pkg/kv/kvserver/replica_rangefeed.go#L607? Can we just get rid of `maybeDisconnectEmptyRangefeed` entirely now?

2.  I found the (existing) shutdown flow hard to reason through. We are calling `p.stopInternal` when the processor registry is already empty here https://github.com/cockroachdb/cockroach/blob/6898797e5784a8d3007a6712666f39c5eaf0c6dd/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L170-L172 . `p.stopInternal` calls `p.reg.DisconnectAllOnShutdown` and `StopProcessor`. `StopProcessor` invokes `p.cleanup` which calls `p.reg.DisconnectAllOnShutdown` once again. It feels like we are repeating this DisconnectAll process multiple times on something that has already been cleared. While this should be idempotent, the repetition felt hard to follow. I wish the flow were clearer or that there were comments that made it easier to justify the correctness of everything.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-:-OCfcEEhAafvxi4vmIjq:bhjlcqu)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @nvanbenschoten, @stevendanna, and @tbg)

___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 74 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCcqbQ565zgkgdWGMXq:-OCfuziqArB7IlQMitcP:b48zbk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/registry.go#L74)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Maybe for a follow-up PR?
</blockquote></details>

Related to the point above - I'm not entirely sure if I'm correct but could we just check for `r.IsDisconnected()` instead of using `setShouldUnregister` and `shouldUnregister`? 
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 297 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCfyrs_CPRRPDATxQ_n:-OCfyrs_CPRRPDATxQ_o:b-5a7xex) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L297)):*
> ```Go
> 
> func (p *ScheduledProcessor) stopInternal(ctx context.Context, pErr *kvpb.Error) {
> 	p.reg.DisconnectAllOnShutdown(ctx, pErr)
> ```

nit: This ties into the comment above - what does moving `p.reg.DisconnectAllOnShutdown` inside `p.stopInternal` buy us? We are now calling `DisconnectAllOnShutdown` when `p.reg.Len()` is already 0 https://github.com/cockroachdb/cockroach/blob/6898797e5784a8d3007a6712666f39c5eaf0c6dd/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L170-L172? There shouldn’t be any new registrations since requests are processed synchronously in the loop and no further registrations will occur after setting `p.stopping` to true. It doesn't seem worth holding up the merge over this. Let's address these nits in follow-up issue https://github.com/cockroachdb/cockroach/issues/135332. 
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 654 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCfwJ3i0LmJXKpWAkly:-OCfwJ3i0LmJXKpWAklz:bwa0c2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eb07d5c2285aa144140ecee11f1120fb1788993/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L654)):*
> ```Go
> 	case <-p.stoppedC:
> 		return
> 	default:
> ```

nit: What does this buy us here? It seems like the request would be simply ignored, and we are not waiting for its completion here.

If this is necessary, should we also check for `p.stopping` https://github.com/cockroachdb/cockroach/blob/6898797e5784a8d3007a6712666f39c5eaf0c6dd/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L650-L655? It doesn't appear to be an issue as it aligns with what `p.enqueueRequest` does https://github.com/cockroachdb/cockroach/blob/6898797e5784a8d3007a6712666f39c5eaf0c6dd/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L635-L641. 


___
*[`pkg/kv/kvserver/rangefeed/buffered_registration.go` line 177 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCffiTOAOaF9cp51fmN:-OCffiTOAOaF9cp51fmO:b-ql3r7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/092e54b2fbc6437875f12280c8dc3fded57d8119/pkg/kv/kvserver/rangefeed/buffered_registration.go#L177)):*
> ```Go
> 		br.mu.disconnected = true
> 		br.stream.SendError(pErr)
> 		br.cleanup(ctx, br)
> ```

The first commit doesn't seem to compile since `ctx` here is not properly declared. Later commits seem to have fixed this, but it might cause issues when doing git bisect between commits.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 361 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/132857#-OCfuU9iEJ25NFTUYMZy:-OCfuU9jECGuTEr8UWAP:b-t9zdv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/55b65a27f5bc281fab7869de4531512f13045a16/pkg/kv/kvserver/rangefeed/registry.go#L361)):*
> ```Go
> 	reg.DisconnectWithErr(ctx, all, pErr)
> }
> ```

This comment seems outdated now.



_Code quote:_
```Go
// This is different from normal disconnect as registrations won't be able to
// perform Unregister when processor's work loop is already terminated.
// This method will cleanup metrics controlled by registry itself beside posting
// errors to registrations.
// TODO: this should be revisited as part of
// https://github.com/cockroachdb/cockroach/issues/110634
```


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
I just realized I forgot to label several comments as nits. Feel free to address the ones you find valuable and ignore the rest.





<!-- Sent from Reviewable.io -->

stevendanna committed:
rangefeed: simplify setShouldUnregister, add comments

Epic: none
Release note: None

stevendanna commented:
I figured out the test failure here. It looks like test problem. Will fix it today.

stevendanna committed:
kvserver/rangefeed: Clear rather than Close memory budget

Clear is similar to Close but also resets the "used" and "reserved"
variables. This shouldn't really matter, but recent changes mean that
sometimes a budget for a processor might be "closed" before a give
registration using that processor's goroutine has finished
draining. This is OK since the feed budget is protected by a mutex and
allocations releases after close are allowed. However, it did break a
test under stress.

Epic: none
Release note: None

stevendanna commented:
bors r+
